### ISSUE TITLE
lgrmsg (or similarly-named) S3 method?

### ISSUE THREAD

**Issue by mmuurr** [s-fleck/lgr#48]
I routinely use custom vectors (often created with the help of [vctrs](https://vctrs.r-lib.org/reference/index.html)) or custom classes (backed by R6, lists, etc.) and like to pass such objects to lgr via the pattern:
```R
lgr::lgr$info("my msg", x = obj)
```
For non-atomic vectors, lgr's default behavior of describing the object (to save space) is nice, but as part of creating the custom classes, I've usually implemented `format()`, too. In cases where I want to use that method, I'll do something like:
```R
lgr::lgr$info("my msg", x = format(obj))
```
... but this is a bit cumbersome to write, and is non-standard as not _all_ objects should be `format`ted (so some lgr statements have `format`, some don't).

Two ideas on which I'm curious about your thoughts:
1. lgr layouts that look for `format` implementations and use them when present (which I think is a bit error-prone, as it's not obvious what to do when a `format` method would be dispatched via S3-inheritance), or
2. A `lgrmsg` S3 method (which would work well for R6 classes) that, when available, works with the existing standard lgr layouts and opts for that string representation in the lgr statements (and when not present, lgr simply uses the existing logic).

I realize one could build their own Layout implementation to do something like this, but I also think it would be a nice addition to the existing Layout(s) for anyone working with custom classes (and would encourage package authors who generate special classes to opt for lgr as their primary logging system :-).

**Comment by s-fleck** [comment]
So I thought about this a bit, and I think exporting something like a `lgrmsg` generic might be a good idea! `format()` isn't really the right generic because it does not necessarily return a `character` scalar (e.g. `format(iris)`). There is also a `toString()` generic in R that always returns a character scalar, but its a bit obscure and the output is also not always desirable for plaintext logging.

Implementation would be easy - just re-branding and exporting my internal `preview_object()` function. The hardest part is coming up with a good name. It should probably be something more broad than `lgrmsg()` or `format_lgr()`, but `preview_object()` is also not great. Let me think a bit on that :)

**Comment by s-fleck** [comment]
So I was thinking of naming the new generic `string_repr()` (representation), inspired by pythons `repr`, but not to conflict with functions from the R package `repr()`. I'll see that I add it in 2-3 weeks, i'll have a bit of time then.

**Comment by mmuurr** [comment]
I like that idea, and never even knew about the `toString()` S3 generic (so, yeah, a bit obscure for sure)!
Maybe something emphasizing the 'compact' representation that's best suited for log outputs ... something ironically non-compact, like `as_compact_string()` :-p.

Purely out of curiosity, why would you want the name to be more 'broad'? I'm sure a compact string representation of objects is useful outside of logging contexts, but putting `log` or `lgr` or something similar in the name might help folks properly differentiate between the motivating uses of `format()`, `str()`, and the _yet-to-be-named_ compact string.

(Regardless of the name, thanks for roadmapping this, it'll be immensely useful!)

**Comment by s-fleck** [comment]
So I originally wrote the function I'm using as a general utility function in my personal utility package sfmisc (mostly for error messages back then). When I started working on lgr I just decided to reuse it there. So I think it does make sense outside of the lgr package. 

I generally tend do "overthink" function names but here are my two/three cents:
- something like `as_compact_string()` or `as_log_string()` is pretty clear, but as you said maybe a bit verbose. I also would want to avoid the `as_` because that implies a data type conversion in R, while `string_repr()` does more than that. You also wouldn't want a function called `as_plot()` or `as_formatted_object()`.
- I still like `string_repr()` best, but I would prefer a verb. 
- `logify()` would also be pretty clear, but also kinda stupid ;) 
- I try to avoid putting `lgr_` (or package-prefixes in general) in front of functions because im a big proponent of just calling package functions with `::` (like `lgr::without_logging()`). 

So as it's still another week or two till I have time to work on it, maybe something better comes up :)
